File: packages/ruby/examples/complete_example.rb
# frozen_string_literal: true

# Complete Example - NatsPubsub v0.3 API
#
# Demonstrates all major features of the new API:
# - Configuration with presets
# - Topic-based messaging
# - Error handling with ErrorAction
# - Batch publishing
# - Testing with TestHarness

require 'nats_pubsub'
require 'securerandom'

# ============================================================================
# 1. CONFIGURATION
# ============================================================================

# Use presets for quick setup
NatsPubsub.configure do |config|
  NatsPubsub::Core::Presets.production(
    config,
    app_name: 'order-service',
    nats_urls: ENV.fetch('NATS_URLS', 'nats://localhost:4222').split(',')
  )
end

# Or customize manually
NatsPubsub.configure do |config|
  config.app_name = 'order-service'
  config.nats_urls = 'nats://localhost:4222'
  config.env = 'production'
  config.concurrency = 20
  config.max_deliver = 5
  config.ack_wait = 30_000
  config.use_dlq = true
  config.use_outbox = false
  config.use_inbox = false
end

# ============================================================================
# 2. SUBSCRIBERS
# ============================================================================

# Order processing subscriber
class OrderProcessor < NatsPubsub::Subscriber
  subscribe_to 'order.created'

  def handle(message, context)
    puts "Processing order #{message['order_id']}"
    puts "Event ID: #{context.event_id}"
    puts "Trace ID: #{context.trace_id}"
    puts "Delivery attempt: #{context.deliveries}"

    # Validate message
    validate_order!(message)

    # Process the order
    process_order_in_database(message)

    # Publish follow-up events
    NatsPubsub.publish(
      topic: 'order.confirmed',
      message: {
        order_id: message['order_id'],
        confirmed_at: Time.now.iso8601
      },
      trace_id: context.trace_id,
      correlation_id: context.correlation_id
    )

    # Send notification
    NatsPubsub.publish(
      topic: 'notification.send',
      message: {
        user_id: message['customer_id'],
        type: 'email',
        subject: 'Order Confirmed',
        body: "Your order #{message['order_id']} has been confirmed."
      },
      trace_id: context.trace_id
    )
  end

  def on_error(error_context)
    error = error_context.error
    message = error_context.message
    attempt_number = error_context.attempt_number
    max_attempts = error_context.max_attempts

    puts "Error processing order #{message['order_id']}: #{error.message}"
    puts "Attempt #{attempt_number}/#{max_attempts}"

    # Decide action based on error type
    case error
    when ValidationError
      # Validation errors should not retry
      puts 'Invalid order data, sending to DLQ'
      NatsPubsub::Core::ErrorAction::DLQ
    when NetworkError, Timeout::Error
      # Network errors should retry
      if error_context.last_attempt?
        puts 'Max retries exhausted, sending to DLQ'
        NatsPubsub::Core::ErrorAction::DLQ
      else
        puts 'Network error, will retry'
        NatsPubsub::Core::ErrorAction::RETRY
      end
    else
      # Default: retry unless exhausted
      if error_context.last_attempt?
        NatsPubsub::Core::ErrorAction::DLQ
      else
        NatsPubsub::Core::ErrorAction::RETRY
      end
    end
  end

  private

  def validate_order!(message)
    required_fields = %w[order_id customer_id items total_amount status]
    missing = required_fields - message.keys

    raise ValidationError, "Missing required fields: #{missing.join(', ')}" unless missing.empty?
    raise ValidationError, 'Items cannot be empty' if message['items'].empty?
    raise ValidationError, 'Total amount must be positive' unless message['total_amount'].positive?
  end

  def process_order_in_database(order)
    # Simulate database operation
    sleep 0.1
  end
end

# Payment processing subscriber
class PaymentProcessor < NatsPubsub::Subscriber
  subscribe_to 'payment.process'

  def initialize
    super
    @circuit_breaker = create_circuit_breaker
  end

  def handle(message, context)
    # Use circuit breaker for external API call
    result = @circuit_breaker.call do
      call_payment_api(message)
    end

    NatsPubsub.publish(
      topic: 'payment.completed',
      message: {
        payment_id: message['payment_id'],
        result: result
      },
      trace_id: context.trace_id
    )
  rescue CircuitBreakerOpenError => e
    puts "Payment API circuit breaker is OPEN: #{e.message}"
    raise StandardError, 'Payment service temporarily unavailable'
  end

  def on_error(error_context)
    if error_context.error.message.include?('temporarily unavailable')
      # Don't retry immediately when circuit is open
      NatsPubsub::Core::ErrorAction::DLQ
    else
      NatsPubsub::Core::ErrorAction::RETRY
    end
  end

  private

  def create_circuit_breaker
    # Simple circuit breaker implementation
    OpenStruct.new(
      failures: 0,
      threshold: 5,
      state: :closed,
      last_opened_at: nil,
      timeout: 30_000,
      call: lambda do |&block|
        if @state == :open && time_since_opened < @timeout
          raise CircuitBreakerOpenError, 'Circuit breaker is OPEN'
        end

        begin
          result = block.call
          @failures = 0
          @state = :closed
          result
        rescue StandardError => e
          @failures += 1
          if @failures >= @threshold
            @state = :open
            @last_opened_at = Time.now
          end
          raise e
        end
      end
    )
  end

  def call_payment_api(message)
    # Simulate external API call
    sleep 0.2
    { success: true, transaction_id: 'txn-123' }
  end

  def time_since_opened
    return Float::INFINITY if @last_opened_at.nil?

    ((Time.now - @last_opened_at) * 1000).to_i
  end
end

# Notification sender
class NotificationSender < NatsPubsub::Subscriber
  subscribe_to 'notification.send'

  def handle(message, context)
    validate_notification!(message)

    puts "Sending #{message['type']} notification to user #{message['user_id']}"

    case message['type']
    when 'email'
      send_email(message['user_id'], message['subject'], message['body'])
    when 'sms'
      send_sms(message['user_id'], message['body'])
    when 'push'
      send_push_notification(message['user_id'], message['subject'], message['body'])
    end
  end

  def on_error(error_context)
    # Non-critical notifications can be discarded on repeated failures
    if error_context.attempt_number >= 3
      puts 'Notification failed after 3 attempts, discarding'
      NatsPubsub::Core::ErrorAction::DISCARD
    else
      NatsPubsub::Core::ErrorAction::RETRY
    end
  end

  private

  def validate_notification!(message)
    required_fields = %w[user_id type subject body]
    missing = required_fields - message.keys

    raise ValidationError, "Missing required fields: #{missing.join(', ')}" unless missing.empty?

    valid_types = %w[email sms push]
    return if valid_types.include?(message['type'])

    raise ValidationError, "Invalid notification type: #{message['type']}"
  end

  def send_email(user_id, subject, body)
    puts "Email sent to user #{user_id}"
  end

  def send_sms(user_id, body)
    puts "SMS sent to user #{user_id}"
  end

  def send_push_notification(user_id, subject, body)
    puts "Push notification sent to user #{user_id}"
  end
end

# ============================================================================
# 3. PUBLISHING
# ============================================================================

# Simple publish
def publish_order_created(order)
  NatsPubsub.publish(
    topic: 'order.created',
    message: order,
    trace_id: generate_trace_id,
    correlation_id: generate_correlation_id
  )
end

# Batch publishing
def publish_multiple_events
  result = NatsPubsub.batch do |batch|
    batch.add('order.created', {
      order_id: '123',
      customer_id: '456',
      items: [{ product_id: 'P1', quantity: 2, price: 29.99 }],
      total_amount: 59.98,
      status: 'pending',
      created_at: Time.now.iso8601
    })

    batch.add('notification.send', {
      user_id: '456',
      type: 'email',
      subject: 'Order Received',
      body: 'We received your order.'
    })

    batch.add('analytics.event.tracked', {
      event: 'order_created',
      user_id: '456',
      properties: { order_id: '123', total_amount: 59.98 }
    })

    batch.with_options(trace_id: generate_trace_id)
  end.publish

  puts "Batch publish: #{result.success_count} succeeded, #{result.failure_count} failed"

  if result.failures.any?
    puts "Failed publishes: #{result.failures.inspect}"
  end

  result
end

# ============================================================================
# 4. APPLICATION LIFECYCLE
# ============================================================================

def start_application
  # Setup topology (creates streams, consumers)
  NatsPubsub.setup

  # Register subscribers
  NatsPubsub.register_subscriber(OrderProcessor.new)
  NatsPubsub.register_subscriber(PaymentProcessor.new)
  NatsPubsub.register_subscriber(NotificationSender.new)

  # Start consuming messages
  NatsPubsub.start

  puts 'Application started successfully'
rescue StandardError => e
  puts "Failed to start application: #{e.message}"
  exit 1
end

def stop_application
  # Stop consuming messages
  NatsPubsub.stop

  # Disconnect from NATS
  NatsPubsub.disconnect

  puts 'Application stopped gracefully'
rescue StandardError => e
  puts "Error during shutdown: #{e.message}"
end

# ============================================================================
# 5. TESTING
# ============================================================================

def test_order_processing
  harness = NatsPubsub::Testing::TestHarness.new(
    subscribers: [OrderProcessor, NotificationSender],
    inline_mode: true
  )

  begin
    harness.setup

    # Publish a test order
    harness.publish('order.created', {
      order_id: 'test-123',
      customer_id: 'cust-456',
      items: [{ product_id: 'P1', quantity: 1, price: 19.99 }],
      total_amount: 19.99,
      status: 'pending',
      created_at: Time.now.iso8601
    })

    # Wait for subscriber to process
    harness.wait_for_subscriber(OrderProcessor, timeout: 5.0)

    # Assert subscriber was called
    raise 'OrderProcessor was not called' unless harness.subscriber_called?(OrderProcessor)
    raise 'OrderProcessor call count mismatch' unless harness.subscriber_call_count(OrderProcessor) == 1

    # Check follow-up notifications were published
    harness.wait_for_messages('notification.send', count: 1)
    notifications = harness.received('notification.send')
    raise 'Notification not found' unless notifications.size == 1

    # Simulate error scenario
    harness.simulate_error(OrderProcessor, StandardError.new('Database connection failed'))

    harness.publish('order.created', {
      order_id: 'test-456',
      customer_id: 'cust-789',
      items: [{ product_id: 'P2', quantity: 1, price: 29.99 }],
      total_amount: 29.99,
      status: 'pending',
      created_at: Time.now.iso8601
    })

    # Check DLQ message
    harness.wait_for(timeout: 5.0) { harness.dlq_messages.any? }
    dlq_messages = harness.dlq_messages
    raise 'DLQ message not found' unless dlq_messages.size == 1

    puts "DLQ message: #{dlq_messages.first.inspect}"
    puts 'All tests passed!'
  ensure
    harness.cleanup
  end
end

# ============================================================================
# 6. HEALTH CHECKS
# ============================================================================

def check_health
  # Full health check
  health = NatsPubsub.health_check
  puts "Health status: #{health.status}"
  puts "Healthy: #{health.healthy?}"
  puts "Components: #{health.components.keys.join(', ')}"

  if health.unhealthy?
    puts 'System is unhealthy!'
  end

  # Quick health check (faster, less detailed)
  quick_health = NatsPubsub.quick_health_check
  puts "Quick health: #{quick_health.status}"
end

# Sinatra health endpoint example
# require 'sinatra/base'
#
# class HealthApp < Sinatra::Base
#   get '/health', to: NatsPubsub.health_check_middleware
#   get '/health/quick', to: NatsPubsub.quick_health_check_middleware
# end

# Rails health endpoint example
# Rails.application.routes.draw do
#   get '/health', to: NatsPubsub.health_check_middleware
#   get '/health/quick', to: NatsPubsub.quick_health_check_middleware
# end

# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

def generate_trace_id
  "trace-#{Time.now.to_i}-#{SecureRandom.hex(4)}"
end

def generate_correlation_id
  "corr-#{Time.now.to_i}-#{SecureRandom.hex(4)}"
end

# Custom error classes
class ValidationError < StandardError; end
class NetworkError < StandardError; end
class CircuitBreakerOpenError < StandardError; end

# ============================================================================
# MAIN
# ============================================================================

if __FILE__ == $PROGRAM_NAME
  # Handle graceful shutdown
  trap('INT') do
    puts 'Received SIGINT, shutting down...'
    stop_application
    exit 0
  end

  trap('TERM') do
    puts 'Received SIGTERM, shutting down...'
    stop_application
    exit 0
  end

  begin
    start_application

    # Keep the process running
    sleep
  rescue StandardError => e
    puts "Fatal error: #{e.message}"
    puts e.backtrace.join("\n")
    exit 1
  end
end

File: packages/javascript/docs/examples/complete-example.ts
/**
 * Complete Example - NatsPubsub v0.3 API
 *
 * Demonstrates all major features of the new API:
 * - Configuration with presets
 * - Schema validation with Zod
 * - Circuit breaker pattern
 * - Batch publishing
 * - Error handling with ErrorAction
 * - Testing with TestHarness
 */

import {
  NatsPubsub,
  Presets,
  subscriber,
  ErrorAction,
  CircuitBreaker,
  createValidator,
  CommonSchemas,
  z,
  TestHarness,
  type MessageContext,
  type ErrorContext,
} from 'nats-pubsub';

// ============================================================================
// 1. CONFIGURATION
// ============================================================================

// Use presets for quick setup
NatsPubsub.configure(
  Presets.production({
    appName: 'order-service',
    natsUrls: process.env.NATS_URLS!.split(','),
  })
);

// Or customize manually
NatsPubsub.configure({
  appName: 'order-service',
  natsUrls: 'nats://localhost:4222',
  env: 'production',
  concurrency: 20,
  maxDeliver: 5,
  ackWait: 30000,
  useDlq: true,
  useOutbox: false,
  useInbox: false,
});

// ============================================================================
// 2. SCHEMA VALIDATION
// ============================================================================

// Define schemas for your messages
const OrderSchema = z.object({
  orderId: CommonSchemas.uuid,
  customerId: CommonSchemas.uuid,
  items: z.array(
    z.object({
      productId: z.string(),
      quantity: z.number().int().positive(),
      price: z.number().positive(),
    })
  ),
  totalAmount: z.number().positive(),
  status: z.enum(['pending', 'confirmed', 'cancelled']),
  createdAt: z.string().datetime(),
});

const NotificationSchema = z.object({
  userId: CommonSchemas.uuid,
  type: z.enum(['email', 'sms', 'push']),
  subject: z.string().min(1),
  body: z.string(),
  metadata: z.record(z.unknown()).optional(),
});

// Create validators
const orderValidator = createValidator(OrderSchema);
const notificationValidator = createValidator(NotificationSchema);

// ============================================================================
// 3. CIRCUIT BREAKER
// ============================================================================

// Create circuit breaker for external API calls
const paymentApiCircuitBreaker = new CircuitBreaker({
  threshold: 5, // Open after 5 failures
  timeout: 30000, // Stay open for 30 seconds
  halfOpenMaxCalls: 3, // Test with 3 calls in half-open state
});

// ============================================================================
// 4. SUBSCRIBERS
// ============================================================================

// Order processing subscriber with schema validation
const OrderProcessor = subscriber('order.created', {
  schema: OrderSchema,
  retryStrategy: {
    maxAttempts: 5,
    backoff: 'exponential',
    initialDelay: 1000,
    maxDelay: 60000,
  },
  circuitBreaker: {
    enabled: true,
    threshold: 5,
    timeout: 60000,
  },
  handler: async (message, context: MessageContext) => {
    console.log(`Processing order ${message.orderId}`);
    console.log(`Event ID: ${context.eventId}`);
    console.log(`Trace ID: ${context.traceId}`);
    console.log(`Delivery attempt: ${context.deliveries}`);

    // Process the order
    await processOrderInDatabase(message);

    // Publish follow-up events
    await NatsPubsub.publish({
      topic: 'order.confirmed',
      message: {
        orderId: message.orderId,
        confirmedAt: new Date().toISOString(),
      },
      metadata: {
        traceId: context.traceId,
        correlationId: context.correlationId,
      },
    });

    // Send notification
    await NatsPubsub.publish({
      topic: 'notification.send',
      message: {
        userId: message.customerId,
        type: 'email',
        subject: 'Order Confirmed',
        body: `Your order ${message.orderId} has been confirmed.`,
      },
      metadata: {
        traceId: context.traceId,
      },
    });
  },
  onError: async (errorContext: ErrorContext) => {
    const { error, message, context, attemptNumber, maxAttempts } = errorContext;

    console.error(`Error processing order ${message.orderId}:`, error);
    console.error(`Attempt ${attemptNumber}/${maxAttempts}`);

    // Decide action based on error type
    if (error.name === 'ValidationError') {
      // Validation errors should not retry
      console.error('Invalid order data, sending to DLQ');
      return ErrorAction.DLQ;
    }

    if (error.name === 'NetworkError' || error.message.includes('timeout')) {
      // Network errors should retry
      if (attemptNumber < maxAttempts) {
        console.log('Network error, will retry');
        return ErrorAction.RETRY;
      } else {
        console.error('Max retries exhausted, sending to DLQ');
        return ErrorAction.DLQ;
      }
    }

    // Default: retry unless exhausted
    if (attemptNumber < maxAttempts) {
      return ErrorAction.RETRY;
    }

    return ErrorAction.DLQ;
  },
});

// Payment processing subscriber with circuit breaker
const PaymentProcessor = subscriber('payment.process', {
  handler: async (message, context: MessageContext) => {
    // Use circuit breaker for external API call
    try {
      const result = await paymentApiCircuitBreaker.execute(async () => {
        return await callPaymentAPI(message);
      });

      await NatsPubsub.publish({
        topic: 'payment.completed',
        message: {
          paymentId: message.paymentId,
          result,
        },
        metadata: {
          traceId: context.traceId,
        },
      });
    } catch (error: any) {
      if (error.name === 'CircuitBreakerError') {
        console.error('Payment API circuit breaker is OPEN');
        // Handle circuit breaker open state
        throw new Error('Payment service temporarily unavailable');
      }
      throw error;
    }
  },
  onError: async (errorContext: ErrorContext) => {
    if (errorContext.error.message.includes('temporarily unavailable')) {
      // Don't retry immediately when circuit is open
      return ErrorAction.DLQ;
    }
    return ErrorAction.RETRY;
  },
});

// Notification sender with validation
const NotificationSender = subscriber('notification.send', {
  schema: NotificationSchema,
  handler: async (message, context: MessageContext) => {
    console.log(`Sending ${message.type} notification to user ${message.userId}`);

    switch (message.type) {
      case 'email':
        await sendEmail(message.userId, message.subject, message.body);
        break;
      case 'sms':
        await sendSMS(message.userId, message.body);
        break;
      case 'push':
        await sendPushNotification(message.userId, message.subject, message.body);
        break;
    }
  },
  onError: async (errorContext: ErrorContext) => {
    // Non-critical notifications can be discarded on repeated failures
    if (errorContext.attemptNumber >= 3) {
      console.warn('Notification failed after 3 attempts, discarding');
      return ErrorAction.DISCARD;
    }
    return ErrorAction.RETRY;
  },
});

// ============================================================================
// 5. PUBLISHING
// ============================================================================

// Simple publish
async function publishOrderCreated(order: z.infer<typeof OrderSchema>) {
  // Validate before publishing
  const validation = orderValidator.validate(order);
  if (!validation.success) {
    throw new Error(`Invalid order: ${JSON.stringify(validation.errors)}`);
  }

  await NatsPubsub.publish({
    topic: 'order.created',
    message: order,
    metadata: {
      traceId: generateTraceId(),
      correlationId: generateCorrelationId(),
    },
  });
}

// Batch publishing
async function publishMultipleEvents() {
  const result = await NatsPubsub.batch()
    .add('order.created', {
      orderId: '123',
      customerId: '456',
      items: [{ productId: 'P1', quantity: 2, price: 29.99 }],
      totalAmount: 59.98,
      status: 'pending',
      createdAt: new Date().toISOString(),
    })
    .add('notification.send', {
      userId: '456',
      type: 'email',
      subject: 'Order Received',
      body: 'We received your order.',
    })
    .add('analytics.event.tracked', {
      event: 'order_created',
      userId: '456',
      properties: { orderId: '123', totalAmount: 59.98 },
    })
    .withOptions({
      traceId: generateTraceId(),
    })
    .publish();

  console.log(`Batch publish: ${result.successCount} succeeded, ${result.failureCount} failed`);

  if (result.failures.length > 0) {
    console.error('Failed publishes:', result.failures);
  }
}

// ============================================================================
// 6. APPLICATION LIFECYCLE
// ============================================================================

async function startApplication() {
  try {
    // Setup topology (creates streams, consumers)
    await NatsPubsub.setup();

    // Register subscribers
    NatsPubsub.subscribeTo(OrderProcessor);
    NatsPubsub.subscribeTo(PaymentProcessor);
    NatsPubsub.subscribeTo(NotificationSender);

    // Start consuming messages
    await NatsPubsub.start();

    console.log('Application started successfully');
  } catch (error) {
    console.error('Failed to start application:', error);
    process.exit(1);
  }
}

async function stopApplication() {
  try {
    // Stop consuming messages
    await NatsPubsub.stop();

    // Disconnect from NATS
    await NatsPubsub.disconnect();

    console.log('Application stopped gracefully');
  } catch (error) {
    console.error('Error during shutdown:', error);
  }
}

// ============================================================================
// 7. TESTING
// ============================================================================

async function testOrderProcessing() {
  const harness = new TestHarness({
    subscribers: [OrderProcessor, NotificationSender],
    inlineMode: true,
  });

  try {
    await harness.setup();

    // Publish a test order
    await harness.publish('order.created', {
      orderId: 'test-123',
      customerId: 'cust-456',
      items: [{ productId: 'P1', quantity: 1, price: 19.99 }],
      totalAmount: 19.99,
      status: 'pending',
      createdAt: new Date().toISOString(),
    });

    // Wait for subscriber to process
    await harness.waitForSubscriber(OrderProcessor, { timeoutMs: 5000 });

    // Assert subscriber was called
    console.assert(harness.subscriberCalled(OrderProcessor));
    console.assert(harness.subscriberCallCount(OrderProcessor) === 1);

    // Check follow-up notifications were published
    await harness.waitForMessages('notification.send', { count: 1 });
    const notifications = harness.received('notification.send');
    console.assert(notifications.length === 1);

    // Simulate error scenario
    harness.simulateError(OrderProcessor, new Error('Database connection failed'));

    await harness.publish('order.created', {
      orderId: 'test-456',
      customerId: 'cust-789',
      items: [{ productId: 'P2', quantity: 1, price: 29.99 }],
      totalAmount: 29.99,
      status: 'pending',
      createdAt: new Date().toISOString(),
    });

    // Check DLQ message
    await harness.waitFor(() => harness.dlqMessages().length > 0, { timeoutMs: 5000 });
    const dlqMessages = harness.dlqMessages();
    console.assert(dlqMessages.length === 1);
    console.log('DLQ message:', dlqMessages[0]);

    console.log('All tests passed!');
  } finally {
    await harness.cleanup();
  }
}

// ============================================================================
// 8. HEALTH CHECKS
// ============================================================================

async function checkHealth() {
  // Full health check
  const health = await NatsPubsub.healthCheck();
  console.log('Health status:', health.status);
  console.log('Components:', health.components);

  if (health.status !== 'healthy') {
    console.error('System is unhealthy!');
  }

  // Quick health check (faster, less detailed)
  const quickHealth = await NatsPubsub.quickHealthCheck();
  console.log('Quick health:', quickHealth.status);
}

// Express middleware for health endpoint
import express from 'express';
const app = express();

app.get('/health', NatsPubsub.healthCheckMiddleware());
app.get('/health/quick', NatsPubsub.quickHealthCheckMiddleware());

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

function generateTraceId(): string {
  return `trace-${Date.now()}-${Math.random().toString(36).substring(7)}`;
}

function generateCorrelationId(): string {
  return `corr-${Date.now()}-${Math.random().toString(36).substring(7)}`;
}

async function processOrderInDatabase(order: any): Promise<void> {
  // Simulate database operation
  await new Promise((resolve) => setTimeout(resolve, 100));
}

async function callPaymentAPI(message: any): Promise<any> {
  // Simulate external API call
  await new Promise((resolve) => setTimeout(resolve, 200));
  return { success: true, transactionId: 'txn-123' };
}

async function sendEmail(userId: string, subject: string, body: string): Promise<void> {
  console.log(`Email sent to user ${userId}`);
}

async function sendSMS(userId: string, body: string): Promise<void> {
  console.log(`SMS sent to user ${userId}`);
}

async function sendPushNotification(userId: string, subject: string, body: string): Promise<void> {
  console.log(`Push notification sent to user ${userId}`);
}

// ============================================================================
// MAIN
// ============================================================================

if (require.main === module) {
  startApplication()
    .then(() => {
      // Handle graceful shutdown
      process.on('SIGINT', async () => {
        console.log('Received SIGINT, shutting down...');
        await stopApplication();
        process.exit(0);
      });

      process.on('SIGTERM', async () => {
        console.log('Received SIGTERM, shutting down...');
        await stopApplication();
        process.exit(0);
      });
    })
    .catch((error) => {
      console.error('Fatal error:', error);
      process.exit(1);
    });
}

export {
  startApplication,
  stopApplication,
  publishOrderCreated,
  publishMultipleEvents,
  testOrderProcessing,
  checkHealth,
};
